Add new `Rails/HasManyOrHasOneDependent` cop

oboxodo opened the pull request:
New Rails cop to force explicit `:dependent` option on `has_many` and
`has_one` associations.

This is for the [related style-guide](https://github.com/bbatsov/rails-style-guide#has_many-has_one-dependent-option).

```ruby
# bad
class User < ActiveRecord::Base
  has_many :comments
  has_one :avatar
end

# good
class User < ActiveRecord::Base
  has_many :comments, dependent: :restrict_with_exception
  has_one :avatar, dependent: :destroy
end
```

**Replace this text with a summary of the changes in your PR.
The more detailed you are, the better.**

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

bbatsov commented:
What about `dependent :delete`?

oboxodo commented:
Not sure I understand the question. What about it?

[The style-guide](https://github.com/bbatsov/rails-style-guide#has_many-has_one-dependent-option) says a dependent option should be defined, but it doesn't provide a specific recommendation as to what option to use. That's left out to the user and I think that's the right thing thing to do.

We could have a whitelist of options but I think that's getting too deep into rails' internals. Rails will complain if the used option is not supported anyway.

mateusg commented:
@oboxodo I think that should not be applicable if the `has_many` association contains the `through` option. That would be redundant, maybe even error prone, e.g.:

```ruby
class Library < ActiveRecord::Base
  has_many :books, dependent: :destroy
  has_many :authors, through: :books # dependent again?
end

class Book < ActiveRecord::Base
  belongs_to :author
end
```

mateusg commented:
@oboxodo Also, I think I wouldn't include `HasOne` in this cop, seems like a thing for another cop (my 2 cents).

bbatsov commented on config/enabled.yml:
```diff
@@ -1645,6 +1645,11 @@ Rails/HasAndBelongsToMany:
   StyleGuide: 'https://github.com/bbatsov/rails-style-guide#has-many-through'
   Enabled: true
 
+Rails/HasManyDependent:
```
The name of the cop is a bit misleading (because it's not clear it handles `has_one` as well).

bbatsov commented on lib/rubocop/cop/rails/has_many_dependent.rb:
```diff
@@ -0,0 +1,47 @@
+module RuboCop
+  module Cop
+    module Rails
+      # This cop looks for has_many or has_one associations that don't specify a
```
Surround `has_many`, `has_one` and `:dependent` with backticks.

bbatsov commented on lib/rubocop/cop/rails/has_many_dependent.rb:
```diff
@@ -0,0 +1,47 @@
+      # @example
+      #   # bad
+      #   class Post < ActiveRecord::Base
+      #     has_many :comments
```
Add an example with `has_one` as well.

bbatsov commented on lib/rubocop/cop/rails/has_many_dependent.rb:
```diff
@@ -0,0 +1,47 @@
+      #     has_many :comments, dependent: :destroy
+      #   end
+      class HasManyDependent < Cop
+        MSG = '`has_many` and `has_one` associations must specify a ' \
```
Tune the message for whatever was actually detected - otherwise it's a bit odd.

bbatsov commented on lib/rubocop/cop/rails/has_many_dependent.rb:
```diff
@@ -0,0 +1,47 @@
+      #   end
+      class HasManyDependent < Cop
+        MSG = '`has_many` and `has_one` associations must specify a ' \
+          '`dependent` option.'.freeze
```
`:dependent`

bbatsov commented on spec/rubocop/cop/rails/has_many_dependent_spec.rb:
```diff
@@ -0,0 +1,28 @@
+      it 'registers an offense when not specifying any options' do
+        inspect_source("#{has_many_or_has_one} :foo")
+
+        expect(cop.messages).to eq([msg])
```
I'd use the new `expect_offense` here.

bbatsov commented:
> Not sure I understand the question. What about it?

I misread something and thought the cop always suggests the use of the destroy strategy.

oboxodo commented:
I think I've addressed all of the comments, except for the one from @mateusg about the `:through` option. The style-guide doesn't mention it at all.

Rails [does support](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) using `:dependent` with `:through` associations, with a caveat:

> If using with the `:through` option, the association on the join model must be a `belongs_to`, and the records which get deleted are the join records, rather than the associated records.

I would agree that specifying a `:dependent` option on `:through` associations might not be something the guide should force.

What do you guys think?

I'm open for collaborations on adding that restriction! I'm going on vacation for a few weeks this weekend so I might not add that myself.

bbatsov commented:
Guess this would do for now! Thanks!

md5 commented:
👎  to having this on `:through` associations

bbatsov commented:
@md5 This was recently fixed on `master`.
